blueocean-core-js/src/js/capability/CapabilityAugmenter.js

```diff
@@ -0,0 +1,144 @@
+/**
+ * Created by cmeyers on 9/8/16.
+ */
+
+const addClass = (clazz, classMap) => {
+    const className = clazz._class;
+
+    if (!classMap[className]) {
+        // eslint-disable-next-line no-param-reassign
+        classMap[className] = [];
+    }
+
+    classMap[className].push(clazz);
+};
+
+const canWalk = (item) => item && (typeof item === 'object' || Array.isArray(item));
+
+const DEFAULT_IGNORED_PROPS = ['_links'];
+const ACTIONS_PROP_NAME = 'actions';
+
+/**
+ * Decorate an object graph with a '_capabilities' property for each object with a valid '_class'
+ * Usage:
+ *      import { capabilityAugmenter } from '@jenkins-cd/blueocean-core-js';
+ *      const augmentCapability = capabilityAugmenter.augmentCapability;
+ *
+ *      fetch(url, fetchOptions)
+ *          .then(data => augmentCapability(data));
+ */
+export class CapabilityAugmenter {
+
+    constructor(capabilityStore) {
+        this._capabilityStore = capabilityStore;
+        this._loggingEnabled = false;
+    }
+
+    /**
+     * Add "_capabilities" data or all objects with a "_class" property.
+     *
+     * @param {object|Array} data
+     * @param {boolean} includeActions whether to fetch capabilities for items in the 'actions' property
+     * @returns {Promise}
+     */
+    augmentCapabilities(data, includeActions = false) {
+        const classMap = this._findClassesInTree(data, includeActions);
+        return this._resolveCapabilities(data, classMap);
+    }
+
+    enablePerfLogging() {
+        this._loggingEnabled = true;
+    }
+
+    /**
+     * Find all of the distinct "_class" values in supplied object.
+     *
+     * @param {object|Array} data
+     * @param {boolean} includeActions whether to fetch capabilities for items in the 'actions' property
+     * @returns {object} key= "_class" name, value= array of all objects of that class.
+     * @private
+     */
+    _findClassesInTree(data, includeActions = false) {
+        const classMap = {};
+        const nodesToWalk = [data];
+        const nodesAlreadyWalked = [];
+        const ignoredProps = DEFAULT_IGNORED_PROPS.slice();
+
+        if (!includeActions) {
+            ignoredProps.push(ACTIONS_PROP_NAME);
+        }
+
+        const started = new Date().getTime();
+
+        let node = nodesToWalk.shift();
+
+        while (node) {
+            nodesAlreadyWalked.push(node);
```

Comment:
Wouldn't really need this in a JSON response since there's no circular data references, but probably good to do anyway if it was reused in a different context.